### PR TITLE
Add option to reduce context window

### PR DESCRIPTION
**Pull request by drbarq**
This PR adds functionality to allow users to set a lower context window than their LLMs maximum context window size. This is useful when the models performance degrades significantly with larger context windows, allowing users to optimize the tradeoff between context length and model performance.

### Background
With PR #4977 being merged in version 0.14, we now support dynamic context window sizes. This PR builds on that by allowing users to manually set a lower context window size than their LLMs maximum, which can be beneficial in cases where:
- The model shows performance degradation with very large contexts
- Users want to optimize for speed vs context length
- Memory or computational resources need to be conserved

### Changes
- Add token count checking before adding new events to prevent exceeding the window
- Implement truncation logic when token count would exceed the configured limit
- Improve handling of first user message preservation to maintain conversation coherence
- Add comprehensive test case for context window parameter truncation

### Configuration
Users can set `max_input_tokens` in two ways:

1. Through config.toml:
```toml
[llm]
max_input_tokens = 20000
```

2. Through environment variables:
```bash
export LLM_MAX_INPUT_TOKENS=20000
```

### Implementation Details
- Token count is checked before adding each new event
- If adding an event would exceed the context window:
  1. The history is truncated using `_apply_conversation_window`
  2. Action-observation pairs are kept together
  3. The first user message is always preserved
- The truncation is done in a way that maintains conversation coherence

### Testing
Added new test case `test_context_window_parameter_truncation` that verifies:
- Token count checking works correctly
- Truncation occurs when limit would be exceeded
- First user message is preserved
- Action-observation pairs stay together

This implements and enhances the changes from PR #5079.

### CODE REVIEW

**Comment by enyst** [comment]
Thank you for taking this up, @drbarq ! For the record, the reason for this PR "switch" is [here](https://openhands-ai.slack.com/archives/C06U8UTKSAD/p1732235490359809).

File: openhands/controller/agent_controller.py [hunk 1]
```diff
@@ -223,6 +223,21 @@ async def on_event(self, event: Event):
 
         # if the event is not filtered out, add it to the history
         if not any(isinstance(event, filter_type) for filter_type in self.filter_out):
+            # Check if adding this event would exceed context window
+            if self.agent.llm.config.max_input_tokens is not None:
+                # Create temporary history with new event
+                temp_history = self.state.history + [event]
+                try:
+                    token_count = self.agent.llm.get_token_count(temp_history)
```
Comment (enyst) [inline openhands/controller/agent_controller.py:231]: I'm afraid this line doesn't really do what we want. We need to do the token counting on the messages that are sent to the LLM API. OpenHands works with events, and these are events, which it then 'translates' into messages and sends to the LLM API. You can see, as you try to make it work on your machine for your use case, that the number of tokens will not match, some events differ more than others but they all differ. 😅 

An alternative is to define a custom exception like, let's say, TokenLimitExceeded, [here](https://github.com/All-Hands-AI/OpenHands/blob/main/openhands/core/exceptions.py), and move this check to the [LLM class](https://github.com/All-Hands-AI/OpenHands/blob/83add629917de56f4f7a238df36977b6a88bbcfd/openhands/llm/llm.py#L190), then raise the exception when the token comparison fails. Then maybe the exception can be treated like this [ContextWindowExceededError](https://github.com/All-Hands-AI/OpenHands/blob/83add629917de56f4f7a238df36977b6a88bbcfd/openhands/controller/agent_controller.py#L502) is. What do you think, does that make sense?

Comment (drbarq) [reply openhands/controller/agent_controller.py:231]: ahh I see I think I understand.  I was getting unpredictable behavior last night and that makes more sense and this definitely doesn't work.  I'll take another crack at it when I get a chance today.  

I do have a dumb question but I can't recreate it predictably.  How is the config.toml file loaded?  Is it only with certain commands? I can't tell if I broke something on my machine or I am assuming incorrect behavior

Comment (enyst) [reply openhands/controller/agent_controller.py:231]: Let's figure it out! There is something a bit unpredictable about it: when you set values there, then run either with UI or with `main.py` or with `cli.py`, then config.toml is loaded. However, if running with UI, any settings that are defined in the UI (in the Settings window) will _override_ the toml values.
Could that be what happens?

Comment (enyst) [reply openhands/controller/agent_controller.py:231]: And it's absolutely a good question! We're still trying to document it properly, the simple thing that it turns out a bit difficult does say that something's wrong and we should rethink some of it.
For reference: https://github.com/All-Hands-AI/OpenHands/issues/3220

**Comment by mamoodi** [comment]
Hi @drbarq Just wanted to check if this PR is still on your radar?

**Comment by drbarq** [comment]
> Hi @drbarq Just wanted to check if this PR is still on your radar?

@mamoodi no I wasn't able to get it working predictably and I ran out of time sorry, do we want me to close?

**Comment by mamoodi** [comment]
@drbarq if you don't think you will come back to it soon, you can close it and open it anytime you are ready. Keeps the PR queue clean.

**Pull request closed without merging**